### PR TITLE
Add column sanitization checks in `CUDF_TEST_EXPECT_COLUMN_*` macros

### PR DESCRIPTION
**Pull request by SurajAralihalli**
This PR addresses Issue #[12786](https://github.com/rapidsai/cudf/issues/12786)

The listed functions have been modified to incorporate a column sanitization check; otherwise, they will raise a `std::invalid_argument` error.
- `expect_column_properties_equal`
- `expect_column_properties_equivalent`
- `expect_columns_equal`
- `expect_columns_equivalent`

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by SurajAralihalli** [comment]
During local testing, the following tests did not pass:

1. REDUCTIONS_TEST
   - SegmentedReductionStringTest.MaxIncludeNulls

2. CLAMP_TEST
   - ClampStringTest.WithNullableColumn
   - ClampStringTest.WithNullableColumnNullHigh
   - ClampStringTest.WithReplaceString
   - ClampDictionaryTest.WithNullableColumn

3. COPYING_TEST
   - ...32 tests (specific tests not listed)

4. UTILITIES_TEST
   - ColumnUtilitiesListsTest.UnsanitaryLists

5. ROLLING_TEST
   - RollingDictionaryTest.LeadLag

6. JSON_PATH_TEST
   - JsonPathTests.GetJsonObjectInvalidQuery

7. DICTIONARY_TEST
   - DictionarySetKeysTest.StringsKeys
   - DictionarySliceTest.SliceColumn

**Comment by SurajAralihalli** [comment]
It appears that the functions tested in the mentioned tests do not seem to produce sanitized columns. Each function needs a closer examination.

I started by analyzing `ClampStringTest.WithNullableColumn`. Upon manual comparison:

Returned:
- Column: B, b, c, NULL, e, F, G, H, NULL, e, B
- Offsets: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11
- Nullmask: 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1

Expected:
- Column: B, b, c, NULL, e, F, G, H, NULL, e, B
- Offsets: 0, 1, 2, 3, 3, 4, 5, 6, 7, 7, 8, 9
- Nullmask: 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1

I suspect this discrepancy could be because the offsets are computed without considering the null mask. See 
- [clamp.cu : form_offsets_and_char_column()](https://github.com/SurajAralihalli/cudf/blob/7f83e83739863736869b6d1a45ff11e705a23af4/cpp/src/replace/clamp.cu#L51)
- [sizes_to_offsets_iterator.cuh : sizes_to_offsets()](https://github.com/SurajAralihalli/cudf/blob/7f83e83739863736869b6d1a45ff11e705a23af4/cpp/include/cudf/detail/sizes_to_offsets_iterator.cuh#L258)

File: cpp/tests/utilities/column_utilities.cu [hunk 1]
```diff
@@ -804,6 +805,12 @@ bool expect_column_properties_equal(column_view const& lhs,
                                     column_view const& rhs,
                                     debug_output_level verbosity)
 {
+  if (cudf::detail::has_nonempty_nulls(lhs, cudf::get_default_stream())) {
+    throw std::invalid_argument("Unsanitized lhs column");
```
Comment (ttnghia) [inline cpp/tests/utilities/column_utilities.cu:809]: Let's change the message to something like `"lhs column has non-empty nulls"` so similar, as `"unsanitized"` is not directly stating the issue.

**Comment by davidwendt** [comment]
Thanks for finding these. I can look into the strings and dictionary failures.

**Comment by vyasr** [comment]
Thanks! This should really help us tighten up our null handling throughout libcudf.

File: cpp/tests/utilities/column_utilities.cu (hunk 1 above)
Comment (SurajAralihalli) [reply cpp/tests/utilities/column_utilities.cu]: Sure @ttnghia, thanks for the feedback!

File: cpp/tests/utilities/column_utilities.cu [hunk 2]
```diff
@@ -822,6 +829,12 @@ bool expect_column_properties_equivalent(column_view const& lhs,
                                          column_view const& rhs,
                                          debug_output_level verbosity)
 {
+  if (cudf::detail::has_nonempty_nulls(lhs, cudf::get_default_stream())) {
+    throw std::invalid_argument("Unsanitized lhs column");
+  }
+  if (cudf::detail::has_nonempty_nulls(rhs, cudf::get_default_stream())) {
+    throw std::invalid_argument("Unsanitized rhs column");
+  }
```
[the remark this answers is not in the document]
Comment (SurajAralihalli) [reply cpp/tests/utilities/column_utilities.cu]: Addressed!

**Review by ttnghia — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]

**Comment by davidwendt** [comment]
The `UTILITIES_TEST` `ColumnUtilitiesListsTest.UnsanitaryLists` is specifically coded in opposition to this change. It checks that an unsanitized list column matches its sanitized partner. If this PR is accepted, then this test would need to be removed.
https://github.com/rapidsai/cudf/blob/21c90d6a264ee4334084f513b34425d5fdd032f8/cpp/tests/utilities_tests/column_utilities_tests.cpp#L284

**Comment by vyasr** [comment]
> The `UTILITIES_TEST` `ColumnUtilitiesListsTest.UnsanitaryLists` is specifically coded in opposition to this change. It checks that an unsanitized list column matches its sanitized partner. If this PR is accepted, then this test would need to be removed.
> 
> https://github.com/rapidsai/cudf/blob/21c90d6a264ee4334084f513b34425d5fdd032f8/cpp/tests/utilities_tests/column_utilities_tests.cpp#L284

I think that's fine. We did something similar in https://github.com/rapidsai/cudf/pull/14363/commits/25ebec79b6975dfddd5256c3c9ac5ba842f8ca93 from https://github.com/rapidsai/cudf/pull/14363, and we discussed that in https://nvidia.slack.com/archives/C01CW5L51QC/p1699980706006349. Since our policies prohibit libcudf APIs from producing unsanitized outputs, we don't need to test unsanitized inputs. It's up to the user not to construct such inputs if they're creating their inputs manually.

**Comment by davidwendt** [comment]
@SurajAralihalli I would recommend you remove this test https://github.com/rapidsai/cudf/blob/21c90d6a264ee4334084f513b34425d5fdd032f8/cpp/tests/utilities_tests/column_utilities_tests.cpp#L284
in this PR. 
And if you merge with the latest branch-24.02, I expect this PR should then pass CI.

**Comment by SurajAralihalli** [comment]
> @SurajAralihalli I would recommend you remove this test https://github.com/rapidsai/cudf/blob/21c90d6a264ee4334084f513b34425d5fdd032f8/cpp/tests/utilities_tests/column_utilities_tests.cpp#L284
> 
> in this PR. 
> 
> And if you merge with the latest branch-24.02, I expect this PR should then pass CI.

Sure @davidwendt! I'll do that by the end of this week when I get back from my trip. Thanks so much for letting me know.

**Pull request merged**